package.json

```diff
@@ -5,7 +5,9 @@
 	"types": "dist/index.d.ts",
 	"scripts": {
 		"build": "bunx tsc",
-		"prepublishOnly": "bun run build",
+		"changeset": "changeset",
+		"prerelease": "pnpm build",
```

Comment:
Thank you so much for this! This is the type of stuff i am totally blank on.

If i understand it correctly + some questions:
1. Make changes (and commit??)
2. Create a changeset. What does this do? Updates the package.json with the correct version and add some metadata that's stored in `.changeset`?
3. If i push this to main, it creates an additional PR to main? What does this PR do?
4. Will changeset only publish when a release-PR is merged? Or every time i push to main?

Probably some dumb questions here😅

Is there a reason why we need pnpm for building, and not use bun?

Reply:
The changesets that area created are just markdown files with some frontmatter describing the changes. They will typically look something like this:

`.changeset/real-paws-leave.md` (file names are randomized)
```md
---
fallback-email: minor
---

Add MailGrid adapter 
```

Creating a changeset does not bump the version in `package.json`. When you push a changeset to main the `release.yml` action will either update or create a release PR. This release PR simply:

1. Bumps the version in the packages `package.json` according to the changesets in the repository.
    - For instance, if there is a patch changeset and a minor changeset the PR will be bumped to the next minor.
2. Creates/appends the packages `CHANGELOG.md` with this versions release notes (the changesets content)
3. Deletes all changesets

So the PR doesn't really do a release, it just does the dirty work of creating release notes and bumping versions for you.

Whenever something is pushed to main, the `release.yml` action tries to publish the package to npm. If the version already exists on npm it will not be able to publish, so nothing happens. However, if the version does not exist, for instance after merging a release PR that has bumped your version, the GitHub action will publish the new version to npm.

So it doesn't really matter how you add the changesets. Whenever a changeset is pushed to main it will be picked up by the action. You could also omit adding changesets, and instead manually bump the `package.json`. This will also lead to a release being made, however using changesets will generate release notes for you, which is nice 😄 These release notes are also the same release notes dependabot will be displaying in its PRs.

>Is there a reason why we need pnpm for building, and not use bun?

Woops! Good catch! Just me being stupid 😅 I copied the setup from another project I've been working on, [domstolene/designsystem](https://github.com/domstolene/designsystem). I pushed a fix for this now, using bun instead 😄 

Here's an example of how a release PR looks like: https://github.com/domstolene/designsystem/pull/1155

I also recommend adding the [changesets bot](https://github.com/apps/changeset-bot) to the repository. It will remind you to make changesets, and show you how changesets will bump the version, in PRs.

And, finally... only changes that needs to be released needs changesets. Changes to CI or similar that don't affect the users are fine to not create changesets for 😊 

Hope this clears things up! If you have any other questions I'm happy to answer them for you 😄